narwhal/consensus/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::bullshark::Bullshark;
+use crate::consensus::ConsensusProtocol;
+use crate::consensus::ConsensusState;
+use crate::consensus_utils::make_consensus_store;
+use crate::metrics::ConsensusMetrics;
+use config::{Committee, Stake};
+use crypto::PublicKey;
+use fastcrypto::hash::Hash;
+use fastcrypto::hash::HashFunction;
+use prometheus::Registry;
+use rand::distributions::Bernoulli;
+use rand::distributions::Distribution;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use rand::Rng;
+use rand::SeedableRng;
+use std::collections::{BTreeSet, HashMap, HashSet, VecDeque};
+use std::num::NonZeroUsize;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+use test_utils::mock_certificate_with_rand;
+use test_utils::CommitteeFixture;
+#[allow(unused_imports)]
+use tokio::sync::mpsc::channel;
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% change of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
+
+    // The percentage of slow nodes we want to introduce to our sample. Basically a slow node is one
+    // that might be able to produce certificates, but those are never get referenced by others. That has
+    // as an effect that when they are leaders might also not get enough support - or no support at all.
+    // For example, a value of 0.2 means that we want up to 20% of our nodes to behave as slow nodes.
+    pub slow_nodes_percentage: f64,
+
+    // The probability of failing to include a slow node certificate to from the certificates of next
+    // round. For example a value of 0.1 means that 10% of the time fail get referenced by the
+    // certificates of the next round.
+    pub slow_nodes_failure_probability: f64,
+}
+
+struct ExecutionPlan {
+    certificates: Vec<Certificate>,
+}
+
+impl ExecutionPlan {
+    fn hash(&self) -> [u8; crypto::DIGEST_LENGTH] {
+        let mut hasher = crypto::DefaultHashFunction::new();
+        self.certificates.iter().for_each(|c| {
+            hasher.update(c.digest());
+        });
+        hasher.finalize().into()
+    }
+}
+
+#[tokio::test]
```

Comment:
I have checked with @mlogan that simtests have deterministic random number generations. We can leverage the simtest infrastructure for specifying a seed to produce deterministic results, and starting multiple runs to search for a failing seed. Do you think we should use simtest for the randomized tests here?

Another advantage of using simtest is that it would use the real leader schedule instead of the fake schedule for unit tests.

Reply:
Yeah, plus that way you can take advantage of other simtest infrastructure, such as the ability to iterate tests, provide seeds on the command line, and the test can become part of the simtest nightly run.

Reply:
@mwtian good suggestion - to be honest I primarily excluded using simtest for the time being as we don't support them for Narwhal + I wanted to better control the seed as all the tests run sequentially (but that might be equally easy with simtets - not aware). @mystenmark let's catch up on what it would take to use simtests for this.